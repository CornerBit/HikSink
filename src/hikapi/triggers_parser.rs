use minidom::Element;
use serde::{Deserialize, Serialize};

use super::EventIdentifier;

#[derive(Debug, PartialEq, Eq, Deserialize, Serialize, Clone)]
pub struct TriggerItem {
    pub identifier: EventIdentifier,
    pub hik_id: String,
    pub description: String,
}

impl TriggerItem {
    pub fn parse(s: &str) -> Result<Vec<TriggerItem>, TriggerParseError> {
        let root: Element = s.parse()?;
        let event_triggers = root
            .get_child("EventNotification", minidom::NSChoice::Any)
            .unwrap_or(&root)
            .get_child("EventTriggerList", minidom::NSChoice::Any)
            .unwrap_or(&root)
            .children();
        let mut parsed = vec![];

        for event_trigger in event_triggers {
            let hik_id = event_trigger
                .get_child("id", minidom::NSChoice::Any)
                .ok_or_else(|| TriggerParseError::FieldMissing("id".to_string()))?
                .text();
            let event_type = event_trigger
                .get_child("eventType", minidom::NSChoice::Any)
                .ok_or_else(|| TriggerParseError::FieldMissing("eventType".to_string()))?
                .text();
            let description = event_trigger
                .get_child("eventDescription", minidom::NSChoice::Any)
                .map(|e| e.text())
                .unwrap_or_else(String::new);
            let channel = event_trigger
                .get_child("videoInputChannelID", minidom::NSChoice::Any)
                .or_else(|| {
                    event_trigger.get_child("dynVideoInputChannelID", minidom::NSChoice::Any)
                })
                .or_else(|| event_trigger.get_child("inputIOPortID", minidom::NSChoice::Any))
                .or_else(|| event_trigger.get_child("dynInputIOPortID", minidom::NSChoice::Any))
                .map(|e| e.text());

            let event_type = event_type
                .parse()
                .map_err(|e| TriggerParseError::EventTypeInvalid(event_type, e))?;
            let identifier = EventIdentifier::new(channel, event_type);

            parsed.push(TriggerItem {
                hik_id,
                identifier,
                description,
            })
        }

        Ok(parsed)
    }
}

impl From<EventIdentifier> for TriggerItem {
    fn from(e: EventIdentifier) -> Self {
        TriggerItem {
            description: String::new(),
            hik_id: format!(
                "{}{}",
                e.event_type.to_string(),
                e.channel
                    .as_ref()
                    .map(|c| format!("-{}", c))
                    .unwrap_or_default()
            ),
            identifier: e,
        }
    }
}

quick_error! {
    #[derive(Debug)]
    pub enum TriggerParseError {
        XmlInvalid(error: minidom::Error) {
            from()
        }
        ChannelMissing {
            display("Event should contain either channelID or dynChannelID but neither were found")
        }
        FieldMissing(field: String) {
            display("Field was expected but missing: {}", field)
        }
        EventTypeInvalid(provided: String, error: String) {
            display("Event type `{}` was incorrectly formatted: {}", provided, error)
        }
    }
}

#[cfg(test)]
mod test {
    use super::TriggerItem;
    const TRIGGERS_CAM: &str = include_str!("../../samples/triggers_cam.xml");
    const TRIGGERS_NVR: &str = include_str!("../../samples/triggers_nvr.xml");
    const TRIGGERS_PTZ: &str = include_str!("../../samples/triggers_ptz.xml");

    #[test]
    fn test_parse_camera_samples() {
        let parsed = TriggerItem::parse(TRIGGERS_CAM).unwrap();
        insta::assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_parse_nvr_samples() {
        let parsed = TriggerItem::parse(TRIGGERS_NVR).unwrap();
        insta::assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_parse_ptz_samples() {
        let parsed = TriggerItem::parse(TRIGGERS_PTZ).unwrap();
        insta::assert_yaml_snapshot!(parsed);
    }
}
