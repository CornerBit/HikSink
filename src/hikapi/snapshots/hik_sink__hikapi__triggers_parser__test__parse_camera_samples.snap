---
source: src/hikapi/triggers_parser.rs
expression: parsed

---
- identifier:
    channel: "1"
    event_type: Motion
  hik_id: VMD-1
  description: VMD Event trigger Information
- identifier:
    channel: "1"
    event_type: Tamper
  hik_id: tamper-1
  description: shelteralarm Event trigger Information
- identifier:
    channel: "1"
    event_type: DiskFull
  hik_id: diskfull
  description: exception Information
- identifier:
    channel: "1"
    event_type: DiskError
  hik_id: diskerror
  description: exception Information
- identifier:
    channel: "1"
    event_type: NicBroken
  hik_id: nicbroken
  description: exception Information
- identifier:
    channel: "1"
    event_type: IpConflict
  hik_id: ipconflict
  description: exception Information
- identifier:
    channel: "1"
    event_type: IllegalAccess
  hik_id: illaccess
  description: exception Information
- identifier:
    channel: "1"
    event_type: LineDetection
  hik_id: linedetection-1
  description: Linedetection Event trigger Information
- identifier:
    channel: "1"
    event_type: FieldDetection
  hik_id: fielddetection-1
  description: fielddetection Event trigger Information
- identifier:
    channel: "1"
    event_type: VideoMismatch
  hik_id: videomismatch
  description: exception Information
- identifier:
    channel: "1"
    event_type: BadVideo
  hik_id: badvideo
  description: exception Information
- identifier:
    channel: "1"
    event_type: FaceDetection
  hik_id: facedetection-1
  description: facedetection Event trigger Information
- identifier:
    channel: "1"
    event_type: UnattendedBaggage
  hik_id: unattendedBaggage-1
  description: UnattendedBaggage Event trigger Information
- identifier:
    channel: "1"
    event_type: AttendedBaggage
  hik_id: attendedBaggage-1
  description: AttendedBaggage Event trigger Information
- identifier:
    channel: "1"
    event_type: StorageDetection
  hik_id: storageDetection-1
  description: storageDetection Event trigger Information
- identifier:
    channel: "1"
    event_type: SceneChangeDetection
  hik_id: scenechangedetection-1
  description: scenechangedetection Event trigger Information
- identifier:
    channel: "1"
    event_type: FaceSnap
  hik_id: faceSnap-1
  description: faceSnap Event trigger Information

